File: src/main/java/org/cimsbioko/search/SearchIndex.java
package org.cimsbioko.search;

import org.apache.lucene.search.IndexSearcher;
import org.apache.lucene.search.SearcherFactory;
import org.apache.lucene.search.SearcherManager;
import org.apache.lucene.store.Directory;
import org.apache.lucene.store.FSDirectory;

import java.io.File;
import java.io.IOException;

import static org.cimsbioko.App.getApp;


public class SearchIndex {

    private static SearchIndex INSTANCE;

    private SearcherManager searcherManager;

    private SearchIndex() {
    }

    public static SearchIndex getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new SearchIndex();
        }
        return INSTANCE;
    }

    public IndexSearcher acquire() throws IOException {
        if (searcherManager == null) {
            File indexFile = new File(getApp().getApplicationContext().getFilesDir(), "search-index");
            Directory indexDir = FSDirectory.open(indexFile);
            searcherManager = new SearcherManager(indexDir, new SearcherFactory());
        } else if (!searcherManager.isSearcherCurrent()) {
            searcherManager.maybeRefresh();
        }
        return searcherManager.acquire();
    }

    public void release(IndexSearcher searcher) throws IOException {
        searcherManager.release(searcher);
    }
}

File: src/main/java/org/cimsbioko/search/SearchIndex.kt
package org.cimsbioko.search

import org.apache.lucene.search.IndexSearcher
import org.apache.lucene.search.SearcherFactory
import org.apache.lucene.search.SearcherManager
import org.apache.lucene.store.Directory
import org.apache.lucene.store.FSDirectory
import org.cimsbioko.App
import java.io.File
import java.io.IOException

object SearchIndex {

    private val searcherManager by lazy {
        val indexFile = File(App.getApp().applicationContext.filesDir, "search-index")
        val indexDir: Directory = FSDirectory.open(indexFile)
        SearcherManager(indexDir, SearcherFactory())
    }

    @Throws(IOException::class)
    fun acquire(): IndexSearcher = with(searcherManager) {
        if (!isSearcherCurrent) {
            maybeRefresh()
        }
        return acquire()
    }


    @Throws(IOException::class)
    fun IndexSearcher.release() {
        searcherManager.release(this)
    }
}